// @aragon/connect-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var ConnectorJson_1 = require("./connections/ConnectorJson");
Object.defineProperty(exports, "ConnectorJson", { enumerable: true, get: function () { return ConnectorJson_1.default; } });
var App_1 = require("./entities/App");
Object.defineProperty(exports, "App", { enumerable: true, get: function () { return App_1.default; } });
var Organization_1 = require("./entities/Organization");
Object.defineProperty(exports, "Organization", { enumerable: true, get: function () { return Organization_1.default; } });
var Permission_1 = require("./entities/Permission");
Object.defineProperty(exports, "Permission", { enumerable: true, get: function () { return Permission_1.default; } });
var Repo_1 = require("./entities/Repo");
Object.defineProperty(exports, "Repo", { enumerable: true, get: function () { return Repo_1.default; } });
var Role_1 = require("./entities/Role");
Object.defineProperty(exports, "Role", { enumerable: true, get: function () { return Role_1.default; } });
__exportStar(require("./utils/index"), exports);
