// @aragon/connect-thegraph/dist/parsers/permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePermissions = void 0;
const connect_core_1 = require("@aragon/connect-core");
function parsePermissions(result, connector) {
    const org = result.data.organization;
    const permissions = org === null || org === void 0 ? void 0 : org.permissions;
    if (!permissions) {
        throw new Error('Unable to parse permissions.');
    }
    const datas = permissions.map((permission) => {
        return {
            appAddress: permission.appAddress,
            allowed: permission.allowed,
            granteeAddress: permission.granteeAddress,
            params: permission.params.map((param) => {
                return {
                    argumentId: param.argumentId,
                    operationType: param.operationType,
                    argumentValue: param.argumentValue,
                };
            }) || [],
            roleHash: permission.roleHash,
        };
    });
    const allowedPermissions = datas.filter((data) => data.allowed);
    return allowedPermissions.map((data) => {
        return new connect_core_1.Permission(data, connector);
    });
}
exports.parsePermissions = parsePermissions;
