// @aragon/connect-thegraph/dist/parsers/repos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseRepo = void 0;
const connect_core_1 = require("@aragon/connect-core");
async function parseRepo(result, connector) {
    var _a, _b, _c, _d, _e, _f;
    const app = result.data.app;
    const repo = app.repo;
    if (!repo) {
        throw new Error('Unable to parse repo.');
    }
    const data = {
        address: repo.address,
        artifact: (_a = repo.lastVersion) === null || _a === void 0 ? void 0 : _a.artifact,
        contentUri: (_b = repo.lastVersion) === null || _b === void 0 ? void 0 : _b.contentUri,
        manifest: (_c = repo.lastVersion) === null || _c === void 0 ? void 0 : _c.manifest,
        name: repo.name,
        registry: (_d = repo.registry) === null || _d === void 0 ? void 0 : _d.name,
        registryAddress: (_f = (_e = app.repo) === null || _e === void 0 ? void 0 : _e.registry) === null || _f === void 0 ? void 0 : _f.address,
    };
    return connect_core_1.Repo.create(data, connector);
}
exports.parseRepo = parseRepo;
